cpp/src/neighbors/ann_mg/ann_mg.cuh

```diff
@@ -0,0 +1,912 @@
+
+#include "../../utils/nccl_helpers.cuh"
+#define NO_NCCL_FORWARD_DECLARATION
+#include <cuvs/neighbors/ann_mg.hpp>
+#undef NO_NCCL_FORWARD_DECLARATION
+#include <cuvs/neighbors/common.hpp>
+
+namespace cuvs::neighbors::mg {
+using namespace cuvs::neighbors;
+using namespace raft;
+
+void check_omp_threads(const int requirements);
+
+template <typename AnnIndexType, typename T, typename IdxT>
+template <typename Accessor>
+void ann_interface<AnnIndexType, T, IdxT>::build(
+  raft::resources const& handle,
+  const cuvs::neighbors::index_params* index_params,
+  raft::mdspan<const T, matrix_extent<int64_t>, row_major, Accessor> index_dataset)
+{
+  if constexpr (std::is_same<AnnIndexType, ivf_flat::index<T, IdxT>>::value) {
+    auto idx = cuvs::neighbors::ivf_flat::build(
+      handle, *static_cast<const ivf_flat::index_params*>(index_params), index_dataset);
+    index_.emplace(std::move(idx));
+  } else if constexpr (std::is_same<AnnIndexType, ivf_pq::index<IdxT>>::value) {
+    auto idx = cuvs::neighbors::ivf_pq::build(
+      handle, *static_cast<const ivf_pq::index_params*>(index_params), index_dataset);
+    index_.emplace(std::move(idx));
+  } else if constexpr (std::is_same<AnnIndexType, cagra::index<T, IdxT>>::value) {
+    cagra::index<T, IdxT> idx(handle);
+    idx = cuvs::neighbors::cagra::build(
+      handle, *static_cast<const cagra::index_params*>(index_params), index_dataset);
+    index_.emplace(std::move(idx));
+  }
+  resource::sync_stream(handle);
+}
+
+template <typename AnnIndexType, typename T, typename IdxT>
+template <typename Accessor1, typename Accessor2>
+void ann_interface<AnnIndexType, T, IdxT>::extend(
+  raft::resources const& handle,
+  raft::mdspan<const T, matrix_extent<int64_t>, row_major, Accessor1> new_vectors,
+  std::optional<raft::mdspan<const IdxT, vector_extent<int64_t>, layout_c_contiguous, Accessor2>>
+    new_indices)
+{
+  if constexpr (std::is_same<AnnIndexType, ivf_flat::index<T, IdxT>>::value) {
+    auto idx = cuvs::neighbors::ivf_flat::extend(handle, new_vectors, new_indices, index_.value());
+    index_.emplace(std::move(idx));
+  } else if constexpr (std::is_same<AnnIndexType, ivf_pq::index<IdxT>>::value) {
+    auto idx = cuvs::neighbors::ivf_pq::extend(handle, new_vectors, new_indices, index_.value());
+    index_.emplace(std::move(idx));
+  } else if constexpr (std::is_same<AnnIndexType, cagra::index<T, IdxT>>::value) {
+    RAFT_FAIL("CAGRA does not implement the extend method");
+  }
+  resource::sync_stream(handle);
+}
+
+template <typename AnnIndexType, typename T, typename IdxT>
+void ann_interface<AnnIndexType, T, IdxT>::search(
+  raft::resources const& handle,
+  const cuvs::neighbors::search_params* search_params,
+  raft::host_matrix_view<const T, int64_t, row_major> h_queries,
+  raft::device_matrix_view<IdxT, int64_t, row_major> d_neighbors,
+  raft::device_matrix_view<float, int64_t, row_major> d_distances) const
+{
+  int64_t n_rows = h_queries.extent(0);
```

Comment:
We should probaly centralize all of these proxy functions in the main `cuvs::neighbors` namespace. I'm not sure the best way to do this with free functions.